Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/CertificateStateHelper.kt

```diff
@@ -130,11 +136,107 @@ fun IncludeCertificateQrcodeCardBinding.bindValidityViews(
+        certificateToggleGroup.isVisible = secondaryCertificate != null
+        primaryCertificateButton.typeface = Typeface.DEFAULT_BOLD
+        secondaryCertificateButton.typeface = Typeface.DEFAULT
+        certificateToggleGroup.setOnCheckedChangeListener { checkedId ->
+            when (checkedId) {
+                R.id.primary_certificate_button -> {
+                    image.loadAny(primaryCertificate.getValidQrCode(Locale.getDefault().language)) {
+                        crossfade(true)
+                        loadingView(qrCodeCard.image, qrCodeCard.progressBar)
+                    }
+                    primaryCertificateButton.typeface = Typeface.DEFAULT_BOLD
+                    secondaryCertificateButton.typeface = Typeface.DEFAULT
+                }
+                R.id.secondary_certificate_button -> {
+                    image.loadAny(secondaryCertificate?.getValidQrCode(Locale.getDefault().language)) {
+                        crossfade(true)
+                        loadingView(qrCodeCard.image, qrCodeCard.progressBar)
+                    }
+                    primaryCertificateButton.typeface = Typeface.DEFAULT
+                    secondaryCertificateButton.typeface = Typeface.DEFAULT_BOLD
+                }
+            }
+        }
+    }
     certificateBadgeCount.isVisible = badgeCount != 0
     certificateBadgeCount.text = badgeCount.toString()
     certificateBadgeText.isVisible = badgeCount != 0
+
+    when (primaryCertificate.displayedState()) {
+        is CwaCovidCertificate.State.Expired -> {
+            expirationStatusIcon.isVisible = badgeCount == 0
+            (expirationStatusIcon.layoutParams as ConstraintLayout.LayoutParams).verticalBias = 1.0f
+            expirationStatusIcon.setImageDrawable(context.getDrawableCompat(R.drawable.ic_error_outline))
+            expirationStatusText.isVisible = badgeCount == 0
+            expirationStatusText.text = context.getText(R.string.certificate_qr_expired)
+        }
+
+        is CwaCovidCertificate.State.Invalid -> {
+            expirationStatusIcon.isVisible = badgeCount == 0
+            (expirationStatusIcon.layoutParams as ConstraintLayout.LayoutParams).verticalBias = 0f
+            expirationStatusIcon.setImageDrawable(context.getDrawableCompat(R.drawable.ic_error_outline))
+            expirationStatusText.isVisible = badgeCount == 0
+            expirationStatusText.text = context.getText(R.string.certificate_qr_invalid_signature)
+        }
+        is CwaCovidCertificate.State.Blocked -> {
+            expirationStatusIcon.isVisible = badgeCount == 0
+            (expirationStatusIcon.layoutParams as ConstraintLayout.LayoutParams).verticalBias = 0f
+            expirationStatusIcon.setImageDrawable(context.getDrawableCompat(R.drawable.ic_error_outline))
+            expirationStatusText.isVisible = badgeCount == 0
+            expirationStatusText.text = context.getText(R.string.error_dcc_in_blocklist_title)
```

Comment:
The code slightly changes between these three states, maybe extract to a function and parameterize what changes ?
such as:
```kotlin
fun Binding.updateExpirationViews(verticalBias:Float=0f@StringRes statusTextRes:Int){
    expirationStatusIcon.isVisible = badgeCount == 0
    (expirationStatusIcon.layoutParams as ConstraintLayout.LayoutParams).verticalBias = verticalBias
    expirationStatusIcon.setImageDrawable(context.getDrawableCompat(R.drawable.ic_error_outline))
    expirationStatusText.isVisible = badgeCount == 0
    expirationStatusText.text = context.getText(statusTextRes)
}
```

Reply:
Addressed with latest changes